app/controllers/users/sessions_controller.rb

```diff
@@ -13,19 +16,31 @@ def create
   end
 
   def sign_in_with_token
-    user = User.find_by(login_token: params[:login_token])
+    @user.update!(login_token: nil, login_token_valid_until: 1.year.ago)
+    sign_in(@user, scope: :user)
+    redirect_to_dashboard
+  end
+
+  def redirect_from_magic_link
+    @login_token = params[:login_token] if params[:login_token].present?
+  end
 
-    if user.present? && Time.zone.now < user.login_token_valid_until
-      user.update!(login_token: nil, login_token_valid_until: 1.year.ago)
-      sign_in(user, scope: :user)
-      redirect_to helpers.profile_dashboard_url(user)
-    else
+private
+
+  def redirect_to_dashboard
+    redirect_to helpers.profile_dashboard_url(current_user) if current_user.present?
+  end
+
+  def ensure_login_token_valid
+    @user = User.find_by(login_token: params[:login_token])
+
+    if @user.blank? || login_token_expired
       flash[:alert] = "There was an error while logging you in. Please enter your email again."
       redirect_to new_user_session_path
     end
   end
 
-  def redirect_from_magic_link
-    @login_token = params[:login_token] if params[:login_token].present?
+  def login_token_expired
```

Comment:
If it returns a boolean then login_token_expired? would be more conventional and reads a tad better.